crates/starknet-devnet-core/src/blocks/mod.rs

```diff
@@ -86,12 +88,17 @@ impl StarknetBlocks {
         self.hash_to_state.insert(block_hash, state);
     }
 
-    pub fn get_by_block_id(&self, block_id: &BlockId) -> Option<&StarknetBlock> {
+    pub fn get_by_block_id(&self, mut block_id: &BlockId) -> Option<&StarknetBlock> {
+        if !self.blocks_on_demand && block_id == &BlockId::Tag(BlockTag::Pending) {
+            // in normal mode, querying pending block should default to the latest
+            block_id = &BlockId::Tag(BlockTag::Latest);
+        }
+
         match block_id {
             BlockId::Hash(hash) => self.get_by_hash(Felt::from(hash)),
             BlockId::Number(block_number) => self.get_by_num(&BlockNumber(*block_number)),
-            // latest and pending for now will return the latest one
-            BlockId::Tag(_) => {
+            BlockId::Tag(BlockTag::Pending) => Some(&self.pending_block),
```

Comment:
Instead of requiring a mutable `block_id`, I think it's clearer if the `!self.blocks_on_demand` check is done inside the `BlockId::Tag(BlockTag::Pending)` arm of the `match`

Reply:
e8a7f3b87565b0e7e949e89980fe50186bc744c9